src/crypto/hpke/mod.rs

```diff
@@ -0,0 +1,236 @@
+
+            let (encap_key, tag) =
+                single_shot_seal_in_place_detached::<IpaAead, super::IpaKdf, super::IpaKem, _>(
+                    &OpModeS::Base,
+                    pk_r,
+                    &info,
+                    &mut plaintext,
+                    &[],
+                    &mut self.rng,
+                )
+                .unwrap();
+
+            MatchKeyEncryption {
+                encap_key: <[u8; 32]>::from(encap_key.to_bytes()),
+                ct: Box::new(plaintext),
+                tag: <[u8; 16]>::from(tag.to_bytes()),
+            }
+        }
+
+        pub fn open(
+            &self,
+            key_id: KeyIdentifier,
+            mut enc: MatchKeyEncryption,
+        ) -> Result<XorReplicated, BottomError> {
+            let aad =
+                AssociatedData::new(key_id, self.epoch, &Self::HELPER_ORIGIN, &Self::SITE_ORIGIN)
+                    .unwrap();
+            open_in_place(
+                &self.registry,
+                &enc.encap_key,
+                enc.ct.as_mut(),
+                &enc.tag,
+                &aad,
+            )?;
+
+            Ok(XorReplicated::deserialize(enc.ct.as_ref())?)
+        }
+
+        pub fn advance_epoch(&mut self) {
+            self.epoch += 1;
+        }
+    }
+
+    /// Make sure we obey the spec
+    #[test]
+    fn ipa_info_serialize() {
+        let aad =
+            AssociatedData::new(255, 32767, &HelperOrigin("foo"), &SiteOrigin("bar")).unwrap();
+        assert_eq!(
+            b"private-attributionfoo\0bar\0\xff\x7f\xff",
+            aad.to_bytes().as_ref()
+        );
+    }
+
+    #[test]
+    fn decrypt_happy_case() {
+        let rng = StdRng::from_seed([1_u8; 32]);
+        let mut suite = EncryptionSuite::new(1, rng);
+        let match_key = XorReplicated::new(u64::MAX, u64::MAX / 2);
+
+        let enc = suite.seal(0, match_key);
+        let r = suite.open(0, enc).unwrap();
+
+        assert_eq!(match_key, r);
+    }
+
+    #[test]
+    fn decrypt_wrong_aad() {
+        let rng = StdRng::from_seed([1_u8; 32]);
+        let mut suite = EncryptionSuite::new(1, rng);
+        let match_key = XorReplicated::new(u64::MAX, u64::MAX / 2);
+        let enc = suite.seal(0, match_key);
+        suite.advance_epoch();
+
+        let _ = suite.open(0, enc).unwrap_err();
+    }
+
+    #[test]
+    fn decrypt_wrong_key() {
+        let rng = StdRng::from_seed([1_u8; 32]);
+        let mut suite = EncryptionSuite::new(10, rng);
+        let match_key = XorReplicated::new(u64::MAX, u64::MAX / 2);
+        let enc = suite.seal(0, match_key);
+        let _ = suite.open(1, enc).unwrap_err();
+    }
+
+    mod proptests {
+        use super::*;
+        use proptest::prelude::ProptestConfig;
+
+        proptest::proptest! {
+            #![proptest_config(ProptestConfig::with_cases(50))]
+            #[test]
+            fn arbitrary_corruption(bad_byte in 0..15_usize, bad_bit in 0..7_usize, seed: [u8; 32]) {
```

Comment:
Corrupting the ciphertext and tag separately work here, but not if we change the size of the ciphertext.  I recommend building two tests.  See also above with respect to the info string.

Reply:
do you want to test that keeping the tag intact and incrementing the size of the ciphertext leads to decryption error?

Reply:
No, I'm suggesting that a 64-bit match key produces ciphertext and tag that are both 16 bytes, which allows you to cheat and use the same test to damage either from the same entropy.  A 40-bit match key (10 bytes of ciphertext) won't work the same way.

Reply:
got you. You're right, I cheated because initially I had two tests and they were very slow because of encryption. So I danced around it and used the same ct and tag to validate two separate test cases.

I will build two tests instead

Reply:
after combining ct + tag I no longer need to have two tests to validate corruption, but I built another one that checks corruption in `enc`